### PR TITLE
callysto: fix auth logic for use with oauthenticator v16

### LINKED ISSUE #3235: Fix Callysto's custom authenticator
Callysto has a custom OAuthenticator, and I think it was kind of broken before, but now its entirely broken for all users that isn't listed as `admin_users` :'(

The reason is that oauthenticator 16 is not using the `authenticate` method as it did before, so overriding it has a different effect. I'll work on fixing it now.

### PR DESCRIPTION
**Pull request by consideRatio**
- Closes #3235

Currently deployed to staging and prod as I figured it would help us avoid stressing getting this in.

### CODE REVIEW

File: config/clusters/callysto/common.values.yaml
Comment (consideRatio) [inline config/clusters/callysto/common.values.yaml:90]: Instead of defining a new trait, we rely on the existing on provided under the individual `allowed_idps` entries.

File: config/clusters/callysto/common.values.yaml
Comment (consideRatio) [inline config/clusters/callysto/common.values.yaml:93]: Instead of overriding `authenticate` we have to override `check_allowed` based on how it works now in oauthenticator v16.

File: config/clusters/callysto/common.values.yaml
Comment (GeorgianaElena) [inline config/clusters/callysto/common.values.yaml:110]: Just to be consistent and to make sure this doesn't change [upstream](https://github.com/jupyterhub/oauthenticator/blob/2fa1c4818f7ba3444f3e2e44f8ddb38127553ffd/oauthenticator/cilogon.py#L45) and forget about it here:

```suggestion
                    email = auth_model["auth_state"][self.user_auth_state_key]["email"]
```

**Review by GeorgianaElena — COMMENTED** [review]
Thanks for fixing this @consideRatio!

I am not sure if you are aware of https://github.com/jupyterhub/oauthenticator/issues/547, that I opened right after Yuvi implemented this special case authenticator for Callysto.

What do you think about upstreaming this change to authenticator directly?
I still believe this is something useful to have by default in `oauthenticator`.

Also, we're still not enforcing to define `allowed_domains` only when the claim is `email`. So people can still define an `allowed_idp.allowed_domains` even if their `usernama_claim` is not `email`, which will make `check_allowed` to fail, with a potentially confusing error message.

What do you think?

**Comment by consideRatio** [comment]
> I am not sure if you are aware of https://github.com/jupyterhub/oauthenticator/issues/547, that I opened right after Yuvi implemented this special case authenticator for Callysto.

Thanks for linking that and opening it in the first place! I wasn't aware and had it on my todo list to write such issue :)

> Also, we're still not enforcing to define `allowed_domains` only when the claim is `email`. So people can still define an `allowed_idp.allowed_domains` even if their `usernama_claim` is not `email`, which will make `check_allowed` to fail, with a potentially confusing error message.

Yes, this confusion makes me consider it almost a bugfix to make the breaking change of checking against `email` claim instead of username. I think it may not end up being a breaking change, but I'm not sure... Let's follow up such discussion in the linked issue.

> What do you think about upstreaming this change to authenticator directly?

I'd like to take this step first separately because callysto's deployment is broken currently, and I don't want to make us feel forced to push through something upstream because we have put us in a situation needing that.

**Comment by consideRatio** [comment]
Thank you for reviewing this @GeorgianaElena!!!!

**Review by GeorgianaElena — APPROVED** [review]
> I'd like to take this step first separately because callysto's deployment is broken currently, and I don't want to make us feel forced to push through something upstream because we have put us in a situation needing that.

Agree! Then let's merge this and remove it once we've upstreamed it 🚀 🚀 
Actually, @consideRatio, can you please also add a FIXME comment with a link to the upstream issue before merging, so we know where this is being tracked?

**Comment by consideRatio** [comment]
@GeorgianaElena 100% agree about the comment. I linked to two issues, one about wildcards to handle `*.edu` and one about checking against `email` claim and not the chosen username.

**Pull request merged**
